tidy3d/components/base.py

```diff
@@ -731,7 +731,38 @@ def __eq__(self, other):
         """Define == for two Tidy3DBaseModels."""
         if other is None:
             return False
-        return self._json_string == other._json_string
+
+        def check_equal(dict1: dict, dict2: dict) -> bool:
+            """Check if two dictionaries are equal, with special handlings."""
+            if not dict1.keys() == dict2.keys():
+                return False
+
+            for key in dict1.keys():
+
+                val1 = dict1[key]
+                val2 = dict2[key]
+
+                if isinstance(val1, tuple) or isinstance(val2, tuple):
+                    val1 = dict(zip(range(len(val1)), val1))
+                    val2 = dict(zip(range(len(val2)), val2))
+
+                if isinstance(val1, dict) or isinstance(val2, dict):
+                    are_equal = check_equal(val1, val2)
+                    if not are_equal:
+                        return False
+
+                elif isinstance(val1, np.ndarray) or isinstance(val2, np.ndarray):
+                    if not np.allclose(np.array(val1), np.array(val2)):
```

Comment:
This errors if `val1` and `val2` are arrays of different shapes. There may be other edge cases too (`val1` is array while `val2` is something on which `np.array(val2)` errors? How about:

```
try: 
    are_equal =  np.allclose(np.array(val1), np.array(val2))
except:
    return False

if not are_equal:
    return False
```

Reply:
hm, good point. I actually saw this function recommended on the internet for something related so maybe I'll do

```py

```

Reply:
Seems to work.

Reply:
[np.array_equal()](https://numpy.org/doc/stable/reference/generated/numpy.array_equal.html)

Reply:
Thanks @ianwilliamson, that's a much better way to do it.